### DISCUSSION TITLE
Custom mapping from Enum to String

### DISCUSSION THREAD

**Discussion by jacekprucia** [mapstruct/mapstruct#3372] [Q&A] [answered]
It appears that when you map from Enum to String, MapStruct just calls `name()` on Enum value. Is there a way to customize that with a method?

I do not really want to use a bunch of `@ValueMapping`, because the mapping of Enum value to String is already stored elsewhere and I simply do not want to duplicate. Plus as source Enum is quite large and it will most surely be cumbersome to maintain. I just would like to introduce custom method `String customMap(MyEnum myEnum);` and just do the mapping myself. Is that possible?

**Comment by filiphr** [comment]
If you want to have a custom mapping between `Enum` and `String` then just provide that custom mapping.

e.g.
```java
default String customMap(MyEnum myEnum) {
    // provide your own custom mapping
}
```

**Comment by jacekprucia** [comment]
I did, but is not being picked up. However I think I need to provide some context, because it might be necessary in this particular scenario. Lets start with objects involved in mapping (getters/setters ommited for brevity) starting with input entity:
```java
public class InputEntity {
  private Long id;
  private Boolean propA;
  private InputEnum propB;
  private String propC;
}
```
...andd then we have output dto:
```java
public class OutputDto {
  private Boolean propA;
  private OutputEnum propB;
  private String propC;
  private String url;
}
```
Just calling `OutputDto map(InputEntity inputEntity)` sets most of the fields, but I want to map `InputEntity` to `url` prop additionally.

Just to make sure I've tried everything I tried using `qualifiedByName` and `@Named`. It worked, but I think I might have idea what is the real source of my problem. To successfully convert `InputEnum` to `String` I need two beans from spring context, namely `MessageSource` and `Locale`. When I use simple methods (without spring beans) everything works fine. When I dabble with `@Context` annotation or try different approaches -- it fails.

What is the correct appropach to injecting beans from spring context, so that I can use them in my custom mapping method? I've been doing it just by declaring my mapper as `public abstract class` and then autowiring beans. This isn't appropriate for this particular scenario as Locale changes with every invocation. I've tried annotating both methods (entity to dto and `Enum` to `String`) with `@Context` but then my custom mapping method stops working.

**Reply by filiphr** [reply]
I am not sure that I understand what you are trying to do. Using `@Context` should work. What stops working? Can you please provide a reproducible repo that is closest to what you have so that we can look into it

**Comment by jacekprucia** [comment]
I've started to extract revelant code into an example repo... and I got it working :) The problem is that my mapper had initially following code:

```
	List<OutputDto> toOutputDtoList(List<InputEntity> inputEntityList);

	@Mapping(source = "propB", target = "description", qualifiedByName = "InputEnumToString")
	OutputDto toOutputDto(InputEntity inputEntity, @Context MessageSource messageSource, @Context Locale locale);
```

...in other words I was expecting mapper to "pick up" both `MessageSource` and `Locale` beans straight from the spring context. This isn't working, unless I add `MessageSource` and `Locale` as explicite arguments from `toOutputDtoList()`. Is there a better way to achieve this? Let me point out that both `MessageSource` and `Locale` are only beans of that particular type available in context.

As per request, example repo is here: https://github.com/jacekprucia/mapstruct-spring

**Reply by filiphr** [reply, marked as answer]
Thanks for the example @jacekprucia. The use of `@Context` is only meant to be passed form the mapping methods, MapStruct is not injecting that in the generated code and thus it is not passing it to the method. What I would suggest to you is to have the mapper inject that and use that.

e.g.
```java
@Mapper(componentModel = "spring")
public abstract class ExampleMapper {

        @Autowired
        private Locale locale;

        @Autowired
        private MessageSource messageSource;

	List<OutputDto> toOutputDtoList(List<InputEntity> inputEntityList);

	@Mapping(source = "propB", target = "description", qualifiedByName = "InputEnumToString")
	OutputDto toOutputDto(InputEntity inputEntity);

	@Named("InputEnumToString")
	protected String inputEnumToString(InputEnum inputEnum) {
		return messageSource.getMessage(inputEnum.name(), new Object[] {}, locale);
	}

}
```

You can also put the `inputEnumToString` method in some other helper service 

e.g.

```java
@Mapper(componentModel = "spring", uses = EnumTranslator.class)
public interface ExampleMapper {

	List<OutputDto> toOutputDtoList(List<InputEntity> inputEntityList);

	@Mapping(source = "propB", target = "description", qualifiedByName = "InputEnumToString")
	OutputDto toOutputDto(InputEntity inputEntity);

}

@Service
public class EnumTranslator {

    protected final Locale locale;
    protected final MessageSource messageSource;
    
    public EnumTranslator(Locale locale, MessageSource messageSource) {
        this.locale = locale;
        this.messageSource = messageSource;
    }

    @Named("InputEnumToString")
    public String inputEnumToString(InputEnum inputEnum, @Context MessageSource messageSource, @Context Locale locale) {
        return messageSource.getMessage(inputEnum.name(), new Object[] {}, locale);
    }

}
```

**Reply by jacekprucia** [reply]
Thank you very much for your answer. This solves all my problems in an elegant way.
